File: java_naif_cli/src/RenderNaive.java
public class RenderNaive {
    private Galaxy galaxie;
    private float[] accelX, accelY, accelZ;
    private final float dt = 0.01f;

    public RenderNaive(Galaxy g) {
        this.galaxie = g;
        accelX = new float[g.size];
        accelY = new float[g.size];
        accelZ = new float[g.size];
    }

    public void execute() {
        for (int i = 0; i < galaxie.size; i++) {
            accelX[i] = accelY[i] = accelZ[i] = 0f;
        }

        for (int i = 0; i < galaxie.size; i++) {
            for (int j = 0; j < galaxie.size; j++) {
                if (i != j) {
                    float dx = galaxie.posX[j] - galaxie.posX[i];
                    float dy = galaxie.posY[j] - galaxie.posY[i];
                    float dz = galaxie.posZ[j] - galaxie.posZ[i];
                    float dij = dx * dx + dy * dy + dz * dz;
                    float d3;
                    if (dij < 1.0f) {
                        d3 = 10.0f * galaxie.mass[j];
                    } else {
                        float sqrtd = (float) Math.sqrt(dij);
                        d3 = 10.0f * galaxie.mass[j] / (sqrtd * sqrtd * sqrtd);
                    }
                    accelX[i] += dx * d3;
                    accelY[i] += dy * d3;
                    accelZ[i] += dz * d3;
                }
            }
        }

        for (int i = 0; i < galaxie.size; i++) {
            galaxie.velX[i] += accelX[i] * 2.0f;
            galaxie.velY[i] += accelY[i] * 2.0f;
            galaxie.velZ[i] += accelZ[i] * 2.0f;
            galaxie.posX[i] += galaxie.velX[i] * dt;
            galaxie.posY[i] += galaxie.velY[i] * dt;
            galaxie.posZ[i] += galaxie.velZ[i] * dt;
        }
    }

    public Galaxy particules() {
        return galaxie;
    }
}

File: Language-naif/rust-naif-cli/src/render_naive.rs
// src/render_naive.rs
use crate::galaxy::Galaxy;

pub struct RenderNaive {
    // En Rust, on sépare souvent les buffers temporaires de la structure de données principale
    // pour éviter les problèmes d'emprunt (borrow checker).
    accel_x: Vec<f32>,
    accel_y: Vec<f32>,
    accel_z: Vec<f32>,
    dt: f32,
}

impl RenderNaive {
    pub fn new(g: &Galaxy) -> Self {
        RenderNaive {
            accel_x: vec![0.0; g.size],
            accel_y: vec![0.0; g.size],
            accel_z: vec![0.0; g.size],
            dt: 0.01,
        }
    }

    pub fn execute(&mut self, galaxie: &mut Galaxy) {
        // bzero equivalent
        self.accel_x.fill(0.0);
        self.accel_y.fill(0.0);
        self.accel_z.fill(0.0);

        let size = galaxie.size;

        // Double boucle O(N^2)
        for i in 0..size {
            for j in 0..size {
                if i != j {
                    let dx = galaxie.pos_x[j] - galaxie.pos_x[i];
                    let dy = galaxie.pos_y[j] - galaxie.pos_y[i];
                    let dz = galaxie.pos_z[j] - galaxie.pos_z[i];

                    let dij = dx * dx + dy * dy + dz * dz;
                    let d3;

                    if dij < 1.0 {
                        d3 = 10.0 * galaxie.mass[j];
                    } else {
                        let sqrtd = dij.sqrt();
                        d3 = 10.0 * galaxie.mass[j] / (sqrtd * sqrtd * sqrtd);
                    }

                    self.accel_x[i] += dx * d3;
                    self.accel_y[i] += dy * d3;
                    self.accel_z[i] += dz * d3;
                }
            }
        }

        // Mise à jour vélocité et position
        for i in 0..size {
            galaxie.vel_x[i] += self.accel_x[i] * 2.0;
            galaxie.vel_y[i] += self.accel_y[i] * 2.0;
            galaxie.vel_z[i] += self.accel_z[i] * 2.0;

            galaxie.pos_x[i] += galaxie.vel_x[i] * self.dt;
            galaxie.pos_y[i] += galaxie.vel_y[i] * self.dt;
            galaxie.pos_z[i] += galaxie.vel_z[i] * self.dt;
        }
    }
}